### DISCUSSION TITLE
uiProgressBar

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#225] [Q&A] [answered]
Has anyone used it, and can give me some tips.

I can do the following
```
pBar = panel:getModulatorByName("modulator-8")
pBar:setValue(0.5,true)
```
However progress shows 0%

if I set value to 1 then progress bar shows 100%, but I can only get 0% or 100%, nothing in between.

interestingly I can do this

```
pBarC = panel:getComponent("modulator-8")
console (String(pBarC:getPropertyDouble("progress")))

```
which returns
0

However if I try to set it to something using

```
pBarC:setPropertyDouble("progress",0.25)
``` 
there is no change.

I can't find anything on the old forum, and I can't really work out anything further from the JUCE documentation.  I suspect the integer isn't getting converted to a double properly in ctrlr

If I can't solve it, I'll build it like Goodweather did in the Pro2 panel, but it's a bit more involved.

**Comment by dnaldoog** [comment, marked as answer]
Here is how I do it: I use timers.

[ProgressBar Example 1.0.zip](https://github.com/RomanKubiak/ctrlr/files/6205716/ProgressBar.Example.1.0.zip)


![progressbar](https://user-images.githubusercontent.com/76831683/112496202-e2eb1e80-8dc7-11eb-81bb-2767718e819a.png)

**Reply by dobo365** [reply]
Yep, same method as I have been using (timers)
But indeed, would be good to have the doc of uiProgress ;-)

**Reply by Godlike-Productions** [reply]
I have it in a timer loop as well. It's to monitor the bulk dumps, which I process through a timer, so as not to overwhelm the poor Z80.  I'll have a look at your example and see if I can adapt it.  Ideally I'd like it in a modal window, but I may just stick it on a new layer.

**Reply by Godlike-Productions** [reply]
I've only just had a chance to check this out.  This example uses the uiProgressBar - thanks for the example.  Trying to implement now.

**Reply by Godlike-Productions** [reply]
Got it implemented.  The uiProgressBar is quite a neat way to do this, though there's still quite a bit of programming around it.

You don't need the for loop in the startProgressBar method though.  The timer will keep retriggering the loop until you stop the timer. That's why it's important to have a stop method inside the callback loop.

**Reply by dobo365** [reply]
When using a timer, you can build it with an if condition to stop it when needed when requested to loop a number of times.
Example to receive a 100 programs dump you would do:

```
if timerid==1 then
	if iCounter<100 then
		RequestProgramDump(iCounter)
		lblProgressPercent:getComponent():setPropertyString("uiLabelText", tostring(math.floor(iCounter/100*100)).." %")
		modProgressBar:getComponent():setValue(iCounter,true)
		iCounter = iCounter + 1
	else 
		lblProgressPercent:getComponent():setPropertyString("uiLabelText", tostring(math.floor(iCounter/100*100)).." %")
		modProgressBar:getComponent():setValue(iCounter, true)
		timer:stopTimer(timerId) 
		iCounter = 0
		lblProgress:getComponent():setVisible(false)
		lblProgressPercent:getComponent():setVisible(false)
		modProgressBar:getComponent():setVisible(false)
		-- Some more actions
	end
end
```
